MINOR: Refactor the MetadataPublisher interface

cmccabe opened the pull request:
This PR refactors MetadataPublisher's interface a bit. There is now a handleControllerChange callback. This is something that some publishers might want. A good example is ZkMigrationClient.

There is now only a single publish() function rather than a separate function for publishing snapshots and log deltas. Most publishers didn't want to do anything different in those two cases. The ones that do want to do something different for snapshots can always check the manifest type.

The close function now has a default empty implementation, since most publishers didn't need to do anything there.

cmccabe committed:
MINOR: Refactor the MetadataPublisher interface

This PR refactors MetadataPublisher's interface a bit. There is now a handleControllerChange
callback. This is something that some publishers might want. A good example is ZkMigrationClient.

There is now only a single publish() function rather than a separate function for publishing
snapshots and log deltas. Most publishers didn't want to do anything different in those two cases.
The ones that do want to do something different for snapshots can always check the manifest type.

The close function now has a default empty implementation, since most publishers didn't need to do
anything there.

rondagostino commented on core/src/main/scala/kafka/server/metadata/DynamicConfigPublisher.scala:
```diff
@@ -31,10 +32,16 @@ class DynamicConfigPublisher(
+  def publish(
+    delta: MetadataDelta,
+    newImage: MetadataImage,
+    manifest: LoaderManifest
```
`Option[LoaderManifest]` since it is optional?

cmccabe replied:
fixed

mumrah commented on core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala:
```diff
@@ -213,7 +213,7 @@ class BrokerMetadataPublisher(
 
       // Apply configuration deltas.
-      dynamicConfigPublisher.publish(delta, newImage)
+      dynamicConfigPublisher.publish(delta, newImage, null)
```
Any way to avoid this null?

cmccabe replied:
fixed

mumrah commented on core/src/main/scala/kafka/server/metadata/DynamicConfigPublisher.scala:
```diff
@@ -31,10 +32,16 @@ class DynamicConfigPublisher(
-  def publish(delta: MetadataDelta, newImage: MetadataImage): Unit = {
+  def name(): String = "DynamicConfigPublisher"
+
+  def publish(
```
Should this have the override annotation?

cmccabe commented:
rebase on trunk

mumrah commented on core/src/main/scala/kafka/server/metadata/DynamicConfigPublisher.scala:
```diff
@@ -35,7 +35,13 @@ class DynamicConfigPublisher(
+  def publish(
+    delta: MetadataDelta,
+    newImage: MetadataImage,
+    manifest: LoaderManifest
```
If we're going to pass the `null` here, we should probably add a comment

mumrah commented on metadata/src/main/java/org/apache/kafka/image/publisher/MetadataPublisher.java:
```diff
@@ -40,33 +40,30 @@ public interface MetadataPublisher extends AutoCloseable {
-            MetadataImage newImage,
-            SnapshotManifest manifest
-    );
+    default void handleControllerChange(LeaderAndEpoch newLeaderAndEpoch) { }
```
nit: "publishControllerChange" to keep the naming consistent?

cmccabe replied:
How about
`void onControllerChange(LeaderAndEpoch newLeaderAndEpoch)`
and
`void onMetadataUpdate(MetadataDelta delta, MetadataImage newImage, LoaderManifest manifest);`?
Might be clearer.

mumrah commented on metadata/src/main/java/org/apache/kafka/image/loader/LoaderManifest.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Contains information about what was loaded.
+ */
+public interface LoaderManifest {
```
Why can't we include the manifest type in the MetadataProvenance?

cmccabe replied:
We need a type to represent the 3-tuple of (offset, epoch, timestamp) since that comes up in many places. I agree that conceptually whether it came from a snapshot or a log is part of the "provenance", but I think it would be ugly in code terms to start putting stuff from the manifest into MetadataProvenance.

mumrah replied:
Ok, fair enough. Works for me 👍

cmccabe committed:
fixes

cmccabe committed:
Merge branch 'trunk' into cmccabe_2023-03-03_refactor_metadatapublisher

cmccabe commented:
I fixed the case where I was invoking `DynamicConfigPublisher.publish` with a null. This was a bit silly... I was just being lazy about adding an extra function.

I also fixed it so that `BrokerMetadataPublisher` is using `DynamicClientQuotaPublisher` and `ScramPublisher`. Ultimately this code should be the same on the broker and controller, just invoked slightly differently. (I have a follow-up change which will use Loader on the broker too, which will simplify this more.)

mumrah commented on metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:
```diff
@@ -449,6 +450,15 @@ SnapshotManifest loadSnapshot(
+                    publisher.onControllerChange(currentLeaderAndEpoch);
+                } catch (Throwable e) {
+                    faultHandler.handleFault("Unhandled error publishing the new leader " +
+                            "change to " + currentLeaderAndEpoch + " with publisher " +
+                            publisher.name(), e);
```
nit; tabs

mumrah commented on metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationDriver.java:
```diff
@@ -223,16 +223,21 @@ public String name() {
+        LoaderManifest manifest
+    ) {
+        enqueueMetadataChangeEvent(delta,
+                newImage,
```
nit: tabs

mumrah commented on metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java:
```diff
@@ -200,7 +201,24 @@ void resetSnapshotCounters() {
+                publishSnapshot(delta, newImage, (SnapshotManifest) manifest);
+                break;
+            default:
+                break;
```
wdyt about throwing an IllegalStateException in the `default`. I do that sometimes to future-proof against unhandled enums down the road.

cmccabe replied:
Actually, let's just remove the `default` case. spotbugs will flag non-exhaustive switch statements on enums. And that's a hard build error for us.

mumrah reviewed (approved):
Thanks @cmccabe, the changes look good. I left a few nitpicks

rondagostino commented on core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala:
```diff
@@ -213,32 +212,13 @@ class BrokerMetadataPublisher(
-            }
-        }
-      }
+      // APply SCRAM delta.
```
nit: `s/AP/Ap/`

rondagostino reviewed (approved):
LGTM

rondagostino commented:
Looks like compile failures in `BrokerMetadataPublisherTest.scala`

cmccabe committed:
fixes

cmccabe committed:
checkstyle
